Ngc 722 fft unzip update (#386)
ska-sa/katgpucbf: doc/fgpu.design.rst

Since the Fourier transform is a linear operator and we defined :math:`z = u + jv`, we also have :math:`Z = U + jV`.

It is important to remember that both :math:`u` and :math:`v` are real-valued, so :math:`U`
and :math:`V` are Hermitian symmetric. By re-arranging things we can reconstruct :math:`U` and
:math:`V` from :math:`Z` using Hermitian symmetry properties. Let :math:`U'`
be :math:`U` with reversed indices i.e., :math:`U'_k = U_{-k}` where indices are taken
modulo :math:`n`.

Hermitian symmetry means that :math:`U'_k = U_{-k} = \overline{U_k}` where
the 'overline' in :math:`\overline{U_k}` denotes conjugation. This is effectively saying that by taking the
reverse indices in :math:`U_k` we get a conjugated result.

Why is this so? Going back to the original definition for the DFT we saw the complex exponential
:math:`e^{\frac{-2\pi j}{2N}\cdot ik}` has a variable :math:`k` where :math:`k` represents the
frequency component under computation for the input sequence :math:`x_i`. If :math:`k` is reversed
(i.e. negative) the complex exponential changes to :math:`e^{\frac{2\pi j}{2n}\cdot i(k)}` as the negaitive

> in :math:`-k` multiplies out.

Reviewer: @james-smith-za do you have an opinion on whether this proof of a standard Fourier result in Fourier is adding anything? I feel like it interrupts the flow of the discussion.

Reviewer 2: I see your point, but I'm not a person who remembers the finer points of mathematical proofs from undergraduate days very well, so I feel as though it does add some value.

Does Sphinx do footnotes? That may be a good compromise.

Moving back to :math:`U` and :math:`V` components then :math:`U' = \overline{U}` and similarly :math:`V' = \overline{V}`.
Going full circle (for completeness) if one was to conjugate the reversal (i.e. :math:`\overline{U'}`) then since
:math:`U' = \overline{U}`, :math:`\overline{U'} = \overline{\overline{U}} = U`. Likewise for

> :math:`\overline{V'} = \overline{\overline{V}} = V`.

Reviewer: @james-smith-za again I think we need a tie-breaker from you - I think stating that conjugating twice cancels out doesn't really add anything.

> Why is this important? Previously we stated that :math:`Z = U + jV`. Now we can consider the reverse (which becomes
> the conjugate as described above) of :math:`Z`, namely :math:`Z'`.

Reviewer: ```suggestion
Why is this important? Previously we stated that :math:`Z = U + jV`. Now we can consider the reverse
of :math:`Z`, namely :math:`Z'`.
```

Z' is **not** the conjugate of Z.

2. It appears to handle real-to-complex transforms by first doing a
   complex-to-complex transform and then using an additional pass to fix up
   the result (i.e. form final FFT output).

For performance reasons, we move part of the Fourier Transform into the
post-processing kernel, and also handle fixing up the real-to-complex transformation.
This is achieved by decomposing the transformation into separately-computed smaller parts
(using the Cooley-Tukey algorithm). Part of the Fourier transform is computed using
cuFFT and the final stage (post-processing kernel) of the process includes one round of
Cooley-Tukey computation and the computation to form the real-to-complex transformation.

To start, let's consider the traditional equation for the Fourier Transform. Let :math:`N`

> be the number of channels into which we wish to decompose the input sequence into, and let

Reviewer: ```suggestion
be the number of channels into which we wish to decompose the input sequence, and let
```

:math:`x_i` be the (real) time-domain samples (:math:`0 \le i < 2N`) and :math:`X_k` be its discrete
Fourier transform (DFT). Because the time domain is real, the frequency domain is Hermitian symmetric,
and we only need to compute half of it to recover all the information. We thus only need to consider
:math:`k` from :math:`0` to :math:`N-1` (this loses information about :math:`X_N`, but it is convenient
to discard it and thus have a power-of-two number of outputs).

.. math:: X_k = \sum_{i=0}^{2N-1} e^{\frac{-2\pi j}{2N}\cdot ik} x_i.

We know that a direct implementation of the DFT is inefficient and alternative, more efficient

> means exist to perform this computation. One such method is the FFT introduce by Cooley-Tukey

Reviewer: ```suggestion
means exist to perform this computation. One such method is the FFT introduced by Cooley-Tukey
```

and in the GPU space cuFFT is one such implementation. As highlighted earlier, transform sizes
of greater than 16384 (for a GeForce RTX 3080 Ti at least) require more than one memory pass making it
less efficient than it needs to be. The technique detailed below uses the decomposition as provided by
Cooley-Tukey to break down a larger transform into smaller 'sub-transforms' where the number of 'sub-transforms'
is intentionally kept small for efficiency reasons and later combined (same process as the FFT) to form
the larger transform size. This is a multi-step process and requires some extra notation and math tricks.

Real-to-complex transform
~~~~~~~~~~~~~~~~~~~~~~~~~

> Now for some notation to see how this works. We start by treating :math:`x` (a real array of length N) as if it is

Reviewer: ```suggestion
Now for some notation to see how this works. We start by treating :math:`x` (a real array of length 2N) as if it is
```

a complex array :math:`z` of length N, with each adjacent pair of real values in :math:`x` interpreted as the
real and imaginary components of a complex value, and computing the Fourier transform of :math:`z`. Formally,
let :math:`u_i = x_{2i}` and :math:`v_i = x_{2i+1}`. Then :math:`z_i = u_i + jv_i = x_{2i} + j x_{2i+1}`.

For clarity, array :math:`x` has length :math:`2N` but is divided into even- and odd- samples
each of length :math:`N`. In other words, each adjacent pair of real values in :math:`x` is interpreted as the real

> and imaginary components of a complex value.

Reviewer: Can delete this paragraph - the previous one now says all this.

We will start by computing the Fourier transform of :math:`z`. This is convenient to do because :math:`z` has the
identical memory layout to :math:`x`, so we can obtain it just by viewing the same memory as containing N complex
numbers instead of 2N real numbers. In other words, each adjacent pair of real values in :math:`x` is interpreted

> as the real and imaginary components of a complex value.

Reviewer: And you can delete everything after the first sentence in this paragraph too (as well as the paragraph break at the end).

Hermitian symmetry means that :math:`U'_k = U_{-k} = \overline{U_k}` where
the 'overline' in :math:`\overline{U_k}` denotes conjugation. This is effectively saying that by taking the

> reverse indices in :math:`U_k` we get a conjugated result (see [#]_ for a reminder why this is the case).

Reviewer: ```suggestion
reverse indices in :math:`U_k` we get a conjugated result (see [#]_ for a reminder of why this is the case).
```

> Looking back at to :math:`U` and :math:`V` components then :math:`U' = \overline{U}` and similarly :math:`V' = \overline{V}`.

Reviewer: Something wrong with that grammar ("Looking back at to").

Unzipping the FFT
~~~~~~~~~~~~~~~~~

> Right - lets get practical and show how we *actually* implement this. From here we'll assume all

Reviewer: Try the suggested change I made and check that it looks sensible in the output.

transforms are complex-to-complex unless specified otherwise. Firstly, some recap: the Cooley-Tukey
algorithm allows a transform of size :math:`c =mn` to be decomposed into :math:`n` transforms of
size :math:`m` followed by :math:`m` transforms of size :math:`n`. We'll refer to :math:`n` as the
"unzipping factor". We will keep it small (typically not more than 4), as the
implementation requires registers proportional to this factor.

To recap the Cooley-Tukey algorithm: let a time-domain index :math:`i` be written as :math:`qn + r`
and a frequency-domain index :math:`k` be written as :math:`pm + s`. Let :math:`z^r` denote the array
:math:`z_r, z_{n+r}, \dots, z_{(m-1)n+r}`, and denote its Fourier transform by :math:`Z^r`. It is
worthwhile to point out that the superscript :math:`r` *does not* denote raise to the power but rather
is a means to indicate an :math:`r^{th}` array. In practice this :math:`r^{th}` array is a subset (part)
of the larger :math:`z` array of input data. Let's unpack this a bit further - What is actually happening

> is that the initial array :math:`z` is divided into :math:`n=4` separate arrays each of

Reviewer: You've added a paragraph break, but you haven't introduced the fact that you're using example values - numbers like 4 and 32768 just appear out of nowhere.

2.4 + 0           8
...               ...
...               ...
8191.4 + 0        32764
==============  ========

... (intentionally skipping 2nd and 3rd array to conserve space)

The fourth array when :math:`r=3` (i.e. :math:`z^{3}`)

==============  ========
   Inputs       Index(i)
--------------  --------
  qn + r          i
==============  ========
0.4 + 3           3
1.4 + 3           7
2.4 + 3           11
...               ...
...               ...
8191.4 + 3        32767
==============  ========

What this shows is that each sub-array consists of samples from the initial array :math:`z` indexed
by :math:`i=qn+r` where each sample is every :math:`4^{th}` and offset by :math:`r`. Pictorially
this looks like,

.. image:: images/z_array_breakdown.png
   :width: 600

Right, so we have separate sub-arrays as indexed form the initial array, what happens next? These various
:math:`z^{r}` arrays are fed to cuFFT yielding 4 complex-to-complex transforms. These separate transforms
now need to be combined to form a single real-to-complex transform of the full initial size.

To see how the :math:`k` indexing works, :math:`k = pm + s` and is dealt with in a similar manner as above.

> The parameter :math:`m=8192` is static (in this case), and :math:`p` has a range :math:`0` to :math:`n-1`

Reviewer: ```suggestion
The parameter :math:`m=8192` is static (in this case), and :math:`p` has a range :math:`0` to :math:`n-1`
```

Not sure what you mean by "static".

implementation requires registers proportional to this factor. We are now going to go step-by-step and
separate the input array :math:`z` into :math:`n` parts of size :math:`m` with each part operated on using a
Fourier transform.

To recap the indexing used in the Cooley-Tukey algorithm: let a
time-domain index :math:`i` be written as :math:`qn + r` and a frequency-domain index :math:`k` be
written as :math:`pm + s`. Let :math:`z^r` denote the array :math:`z_r, z_{n+r}, \dots, z_{(m-1)n+r}`,
and denote its Fourier transform by :math:`Z^r`. It is worthwhile to point out that the superscript
:math:`r` *does not* denote exponentiation but rather is a means to indicate an :math:`r^{th}` array.
In practice this :math:`r^{th}` array is a subset (part) of the larger :math:`z` array of input data.

As a way of an example, let :math:`n=4` ("unzipping factor") and :math:`N=32768` (total number of channels).
Now let's unpack this a bit further --- what is actually happening is that the initial array :math:`z` is divided into
:math:`n=4` separate arrays each of :math:`m=32768/4 = 8192` elements (hence the :math:`N = mn` above).

> This is true for a desired overall transform size of 32768 channels. The actual samples that land up in each

Reviewer: ```suggestion
The actual samples that land up in each
```
You've already said above that the total number of channels is 32768.

Unzipping the FFT
~~~~~~~~~~~~~~~~~
Right --- lets get practical and show how we *actually* implement this. From here we'll assume all
transforms are complex-to-complex unless specified otherwise. Firstly, some recap: the Cooley-Tukey
algorithm allows a transform of size :math:`N = mn` to be decomposed into :math:`n` transforms of
size :math:`m` followed by :math:`m` transforms of size :math:`n`. We'll refer to :math:`n` as the
"unzipping factor". We will keep it small (typically not more than 4), as the
implementation requires registers proportional to this factor. We are now going to go step-by-step and

> separate the input array :math:`x` into :math:`n` parts of size :math:`m` with each part operated on using a

Reviewer: ```suggestion
separate the input array :math:`z` into :math:`n` parts of size :math:`m` with each part operated on using a
```

size :math:`m` followed by :math:`m` transforms of size :math:`n`. We'll refer to :math:`n` as the
"unzipping factor". We will keep it small (typically not more than 4), as the
implementation requires registers proportional to this factor.
implementation requires registers proportional to this factor. We are now going to go step-by-step and
separate the input array :math:`x` into :math:`n` parts of size :math:`m` with each part operated on using a
Fourier transform.

To recap the Cooley-Tukey algorithm: let a time-domain index :math:`i` be
written as :math:`qn + r` and a frequency-domain index :math:`k` be
written as :math:`pm + s`. Let :math:`z^r` denote the array :math:`z_r, z_{n+r},

> \dots, z_{(m-1)n+r}`, and denote its Fourier transform by :math:`Z^r`. Then

Reviewer: By moving these definitions to later on, you're now using them before they're defined.

Author: The definition follows immediately after so is it a major issue?

Author: Reordered to aid the flow.

The first array when :math:`r=0` (i.e. :math:`z^{0}`)

========================   ========
   Inputs                  Index
------------------------   --------
  qn + r                      i
========================   ========
:math:`0 \cdot 4` + 0         0
:math:`1 \cdot 4` + 0         4
:math:`2 \cdot 4` + 0         8
...                           ...
...                           ...
:math:`8191 \cdot 4` + 0    32764
========================   ========

This can be extended to the other remaining arrays. The fourth array when :math:`r=3` (for example), :math:`z^{3}`
is :math:`z_{3}`, :math:`z_{7}`, :math:`z_{11}`, ..., :math:`z_{32767}`.

What this shows is that each sub-array consists of samples from the initial array :math:`z` indexed
by :math:`i=qn+r` where each sample is every :math:`4^{th}` and offset by :math:`r`. Pictorially
this looks like,

> .. image:: images/z_array.png

Reviewer: The image still says "real-valued", should say "complex-valued".

The whole expression is a Fourier transform of the expression in brackets
(the exponential inside the bracket is the so-called "twiddle factor").

Now let's walk through each part. To recap the indexing used in the Cooley-Tukey algorithm: let a
time-domain index :math:`i` be written as :math:`qn + r` and a frequency-domain index :math:`k` be
written as :math:`pm + s`. Let :math:`z^r` denote the array :math:`z_r, z_{n+r}, \dots, z_{(m-1)n+r}`,
and denote its Fourier transform by :math:`Z^r`. It is worthwhile to point out that the superscript

> :math:`r` *does not* denote raise to the power but rather is a means to indicate an :math:`r^{th}` array.

Reviewer: ```suggestion
:math:`r` *does not* denote exponentiation but rather is a means to indicate an :math:`r^{th}` array.
```

In practice this :math:`r^{th}` array is a subset (part) of the larger :math:`z` array of input data.

Let's unpack this a bit further --- what is actually happening
is that the initial array :math:`z` is divided into :math:`n=4` separate arrays each of

> :math:`m=32768/4 = 8192` elements (hence the :math:`N = mn` above)(This is true for a desired overall

Reviewer: Two parenthetical remarks next to each other looks odd and is hard to read.

transform size of 32768 channels). The actual samples that land up in each array are defined by the

> indices :math:`i` and :math:`k`. Lets start with :math:`i`.

Reviewer: ```suggestion
indices :math:`i` and :math:`k`. Let's start with :math:`i`.
```

Reviewer: Ending a paragraph "Let's start with i" is a bit of a cliff-hanger. It might be more sensible to use that as the start of the next paragraph.

It was stated that :math:`i = qn + r`. The parameter :math:`r` takes on the range :math:`0` to :math:`n-1`
(so :math:`r=0` to :math:`r=3` as :math:`n = 4`) and :math:`q` takes on the
range :math:`0` to :math:`m-1` (i.e. :math:`q=0` to :math:`q=8191`).
So we are dividing up array :math:`z` into :math:`n` smaller arrays denoted by  :math:`r` (i.e. :math:`z^{r}`)
each of length :math:`m=8192`. So what does this look like practically?

The first array when :math:`r=0` (i.e. :math:`z^{0}`)

========================   ========
   Inputs                  Index(i)
------------------------   --------

>   qn + r                      i

Reviewer: Not sure what it means to have two headings on a column, one of which is "Index(i)" and the other is "i".

========================   ========
:math:`0 \cdot 4` + 0         0
:math:`1 \cdot 4` + 0         4
:math:`2 \cdot 4` + 0         8
...               ...

> ...               ...

Reviewer: Are the second set of dots in each line meant to be in the second column?

:math:`8191 \cdot 4` + 0    32764
========================   ========

... (intentionally skipping 2nd and 3rd array to conserve space)

The fourth array when :math:`r=3` (i.e. :math:`z^{3}`)

========================  ========
   Inputs                 Index(i)
------------------------  --------
  qn + r                     i
========================  ========
:math:`0 \cdot 4 + 3`        3
:math:`1 \cdot 4 + 3`        7
:math:`2 \cdot 4 + 3`        11
...               ...
...               ...
:math:`8191 \cdot 4 + 3`   32767

> ========================  ========

Reviewer: These tables take up an awful lot of space, particularly since you have a diagram as well. Can one just say something like, "For example, z^0 is z_0, z_4, z_8, ..., z_32764, while z^3 is z_3, z_7, z_11, ..., z_32767"? Or even just delete it and let the diagram speak for itself?

What this shows is that each sub-array consists of samples from the initial array :math:`z` indexed
by :math:`i=qn+r` where each sample is every :math:`4^{th}` and offset by :math:`r`. Pictorially
this looks like,

> .. image:: images/z_array.png

Reviewer: Where is the source for this image?

Reviewer: The image still references x rather than z; and the caption in the middle makes it sound like z^r are the outputs of a Fourier transform, rather than the inputs.

Author: Previous feedback suggested the image be modified to x_0 to x_65535. I'll commit the source (.ppt).

Previous feedback:
The top line refers to z, but it should refer to x, and the values should be labelled x_0 to x_65535, not r_0 to r_65536.
It shows the z^r arrays each containing either only real or only imaginary components. These are complex arrays. At this stage we're not worrying about the real-to-complex transform, and I think the diagram would be less confusing if it only dealt with z and showed complex numbers without trying to split them into real and imaginary.

Author: Image updated as requested.

   :width: 600

> (click image to enlarge)

Reviewer: Sphinx is giving a warning:
```
katgpucbf/doc/fgpu.design.rst:399: WARNING: Explicit markup ends without a blank line; unexpected unindent.
```

It's also not going to make any sense in the PDF output (I think Sphinx has some directives that let you limit some stuff to a particular output format).

I feel like just using a bigger font (and maybe fewer elements) would make it readable without the need to enlarge.

Right, so we have separate sub-arrays as indexed from the initial array, what happens next? These various

> :math:`z^{r}` arrays are fed to cuFFT yielding 4 complex-to-complex transforms. These separate transforms

Reviewer: ```suggestion
:math:`z^{r}` arrays are fed to cuFFT yielding :math:`n` complex-to-complex transforms. These separate transforms
```

now need to be combined to form a single real-to-complex transform of the full initial size. An inconvenience
of this structure is that :math:`z^r` is not a contiguous set of input samples, but a strided array.
While cuFFT does support both strided inputs and batched transformations, we cannot batch over :math:`r`
and over multiple spectra at the same time as it only supports a single batch dimension with corresponding
stride. We solve this by modifying the PFB kernel to reorder its output such that each :math:`z^r` is output
contiguously. This can be done by shuffling some bits in the output index (because we assume powers of two
everywhere).

> To see how the :math:`k` indexing works, :math:`k = pm + s` and is dealt with in a similar manner as above.

Reviewer: Comma splice.

Parameter :math:`m = 8192` (in this example), and :math:`p` has a range :math:`0` to :math:`n-1`
(i.e. :math:`p = 0` to :math:`p = 3` as :math:`n = 4` in our example); and :math:`s` takes on the
range :math:`0` to :math:`m-1` (i.e. :math:`s = 0` to :math:`s = 8191`).

Looking at this practically,

When :math:`p = 0`

===========================  =========
   Inputs                     Index(k)
---------------------------  ---------
  pm + s                      k
===========================  =========
:math:`0 \cdot 8192` + 0      0
:math:`0 \cdot 8192` + 1      1
:math:`0 \cdot 8192` + 2      2
...               ...

> ...               ...

Reviewer: Again, second set of ... seems to be in the wrong column.

:math:`0 \cdot 8192` + 8191   8191
===========================  =========

... (intentionally skipping a few to conserve space)

When :math:`p=3`

===========================  ========
   Inputs                    Index(k)
---------------------------  --------
  pm + s                         k
===========================  ========
:math:`3 \cdot 8192` + 0      24576
:math:`3 \cdot 8192` + 1      24577
:math:`3 \cdot 8192` + 2      24578
...              ...
...              ...
:math:`3 \cdot 8192` + 8191   32767

> ===========================  ========

Reviewer: Again, these tables take a huge amount of space and could be replaced by a couple of sentences.

performance significantly, and not always in obvious ways. It will likely be
necessary to implement autotuning to get optimal results across a range of
problem parameters and hardware devices, but this has not yet been done.

.. rubric:: Footnotes

> .. [#] Going back to the original definition for the DFT we saw the complex exponential :math:`e^{\frac{-2\pi j}{2N}\cdot ik}` has a variable :math:`k` where :math:`k` represents the frequency component under computation for the input sequence :math:`x_i`. If :math:`k` is reversed (i.e. negative) the complex exponential changes to :math:`e^{\frac{2\pi j}{2n}\cdot i(k)}` as the negative in :math:`-k` multiplies out.

Reviewer: Wrap this rather long line - it will make future changes easier to review.

Reviewer: Also it looks a bit odd that this line uses `ik` in the first formula for `i(k)` in the second.

> .. [#] Going back to the original definition for the DFT we saw the complex exponential :math:`e^{\frac{-2\pi j}{2N}\cdot ik}` has a variable :math:`k` where :math:`k` represents the frequency component under computation for the input sequence :math:`x_i`. If :math:`k` is reversed (i.e. negative) the complex exponential changes to :math:`e^{\frac{2\pi j}{2n}\cdot i(k)}` as the negative in :math:`-k` multiplies out.

Reviewer: ```suggestion
.. [#] Going back to the original definition for the DFT we saw the complex exponential :math:`e^{\frac{-2\pi j}{2N}\cdot ik}` has a variable :math:`k` where :math:`k` represents the frequency component under computation for the input sequence :math:`x_i.` If :math:`k` is reversed (i.e. negative) the complex exponential changes to :math:`e^{\frac{2\pi j}{2n}\cdot i(k)}` as the negative in :math:`-k` multiplies out.
```

At the moment the full stop is getting wrapped onto the next line in the HTML output. Putting it inside the math prevents the line being broken in the wrong place.

> .. [#] Going back to the original definition for the DFT we saw the complex exponential :math:`e^{\frac{-2\pi j}{2N}\cdot ik}` has a variable :math:`k` where :math:`k` represents the frequency component under computation for the input sequence :math:`x_i`. If :math:`k` is reversed (i.e. negative) the complex exponential changes to :math:`e^{\frac{2\pi j}{2n}\cdot i(k)}` as the negative in :math:`-k` multiplies out.

Reviewer: ```suggestion
.. [#] Going back to the original definition for the DFT we saw the complex exponential :math:`e^{\frac{-2\pi j}{2N}\cdot ik}` has a variable :math:`k` where :math:`k` represents the frequency component under computation for the input sequence :math:`x_i`. If :math:`k` is reversed (i.e. negative) the complex exponential changes to :math:`e^{\frac{2\pi j}{2N}\cdot i(k)}` as the negative in :math:`-k` multiplies out.
```
